test/test_datapipe.py

```diff
@@ -248,6 +256,128 @@ def test_groupbykey_iterable_datapipe(self):
+        socketserver.TCPServer.allow_reuse_address = True
+
+        server = socketserver.TCPServer(("", 0), Handler)
+        server_addr = "{host}:{port}".format(host=server.server_address[0], port=server.server_address[1])
+        server_thread = threading.Thread(target=server.serve_forever)
+        server_thread.start()
+
+        # Wait a bit for the server to come up
+        time.sleep(3)
+
+        return (server_thread, server_addr, server)
+    except Exception:
+        raise
+
+
+def create_temp_files_for_serving(tmp_dir, file_count, file_size,
+                                  file_url_template):
+    furl_local_file = os.path.join(tmp_dir, "urls_list")
+    with open(furl_local_file, 'w') as fsum:
+        for i in range(0, file_count):
+            f = os.path.join(tmp_dir, "webfile_test_{num}.data".format(num=i))
+
+            write_chunk = 1024 * 1024 * 16
+            rmn_size = file_size
+            while rmn_size > 0:
+                with open(f, 'ab+') as fout:
+                    fout.write(os.urandom(min(rmn_size, write_chunk)))
+                rmn_size = rmn_size - min(rmn_size, write_chunk)
+
+            fsum.write(file_url_template.format(num=i))
+
+
+class TestIterableDataPipeHttp(TestCase):
+    __server_thread: threading.Thread
+    __server_addr: str
+    __server: socketserver.TCPServer
+
+    @classmethod
+    def setUpClass(cls):
+        try:
+            (cls.__server_thread, cls.__server_addr,
+             cls.__server) = setUpLocalServerInThread()
+        except Exception as e:
+            warnings.warn("TestIterableDataPipeHttp could\
+                          not set up due to {0}".format(str(e)))
+
+    @classmethod
+    def tearDownClass(cls):
+        try:
+            cls.__server.shutdown()
+            cls.__server_thread.join(timeout=15)
+        except Exception as e:
+            warnings.warn("TestIterableDataPipeHttp could\
+                           not tear down (clean up temp directory or terminate\
+                           local server) due to {0}".format(str(e)))
+
+    def http_test_base(self, test_file_size, test_file_count, timeout=None,
+                       chunk=None):
+        with tempfile.TemporaryDirectory(dir=os.getcwd()) as tmpdir:
+            # create tmp dir and files for test
+            base_tmp_dir = os.path.basename(os.path.normpath(tmpdir))
+            file_url_template = ("http://{server_addr}/{tmp_dir}/"
+                                 "/webfile_test_{num}.data\n")\
+                .format(server_addr=self.__server_addr, tmp_dir=base_tmp_dir,
+                        num='{num}')
+            create_temp_files_for_serving(tmpdir, test_file_count,
+                                          test_file_size, file_url_template)
+
+            datapipe_dirf = dp.iter.ListDirFiles(tmpdir, '*_list')
+            datapipe_rf = dp.iter.ReadLinesFromFile(datapipe_dirf)
+            datapipe_tup = TupleToElementIDP(datapipe_rf, 1)
```

Comment:
If it's required, can you use map datapipe?

Reply:
sure -- let me quickly change it

Reply:
I replace TupleToElementIDP with MapIterDataPipe while keeping the explicit pipe stack of ListDirFiles and ReadLinesFromFile. My thinking is TupleToElementIDP is more temporary, used specifically for unittest. The other two could follow the pattern of TestIterableDataPipeBasic.

Reply:
It can be simplified as:
```py
def _get_from_tuple(data):
    return data[1]
datapipe_tup = dp.iter.Map(datapipe_f_lines, _get_from_tuple)
```

Reply:
cool. then my proposed change is quite similar with yours

def _get_data_from_tuple_fn(data, *args, **kwargs):
            return data[args[0]]
...
datapipe_line_url = dp.iter.Map(datapipe_f_lines,
                                _get_data_from_tuple_fn, [1])